### PR TITLE
Add `Owl.Data.from_ansidata/1`

### PR DESCRIPTION
**Pull request by zachallaun**
Implements #9. ~~This isn't quite ready to merge, but wanted to open it as a draft for now. The primary things left are to add docs and quite a lot more tests.~~

Notes:

* I've created an internal `Owl.Data.Sequence` module that takes a similar approach to `IO.ANSI`. It's responsible for identifying escape sequences, grouping sequences into types, knowing what the defaults are, etc.
* `maybe_wrap_to_tag`, `reverse_and_tag`, and `collapse_sequences` are unchanged, just moved.
* I'm not making an effort to "combine" tags in any intelligent way at the moment. See the nested red/underlined test for example. Other than for aesthetic purposes, I don't know that there's a real benefit to this and I believe it would make the implementation rather more complicated.

Example:

```elixir
%{foo: 1, bar: "two", baz: 'three'}
|> Inspect.Algebra.to_doc(Inspect.Opts.new(syntax_colors: IO.ANSI.syntax_colors()))
|> Inspect.Algebra.format(80)
|> Owl.Data.from_ansidata()
|> Owl.IO.puts()
```

image

### CODE REVIEW

File: lib/owl/data/sequence.ex [hunk 1]
```diff
@@ -0,0 +1,95 @@
+defmodule Owl.Data.Sequence.Helper do
+  @moduledoc false
+
+  defmacro defsequence_type(name, type) do
+    quote bind_quoted: [name: name, type: type], generated: true do
```
Comment (zachallaun) [inline lib/owl/data/sequence.ex:5]: `generated: true` is necessary here to prevent a "clause cannot match because of ..." warning due to `:inverse` and `:reverse` using the same escape sequence. See b69001f9a7d3689d58678d94f91835f00ce580c1 for the alternative.

Comment (fuelen) [reply lib/owl/data/sequence.ex:5]: I think alternative solution with explicit option is better.

Comment (zachallaun) [reply lib/owl/data/sequence.ex:5]: Okay, cool! I'm going to force-push to remove this last commit, then.

File: lib/owl/data/sequence.ex [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  """
+  def ansi_to_name(binary) when is_binary(binary) do
+    name_by_sequence(binary)
+  end
+
+  @doc """
+  Get the sequence type of an escape sequence or `nil` if not a sequence.
+  """
+  def ansi_to_type(binary) when is_binary(binary) do
+    if name = ansi_to_name(binary) do
```
Comment (fuelen) [inline lib/owl/data/sequence.ex:36]: We should support RGB colors as well

File: lib/owl/data.ex [hunk 3]
```diff
@@ -296,33 +298,64 @@ defmodule Owl.Data do
...
 
+  defp tag_all(data, open_tags) do
+    case Map.values(open_tags) do
+      [] -> data
+      tags -> tag(data, tags)
+    end
+  end
+
+  defp update_open_tags(open_tags, type, name) do
+    if name == Sequence.default_value_by_type!(type) do
```
Comment (fuelen) [inline lib/owl/data.ex:348]: seems like default_value_by_type! cannot return nil

File: lib/owl/data/sequence.ex (hunk 2 above)
Comment (zachallaun) [reply lib/owl/data/sequence.ex:36]: I _believe_ this is already supported, but working on adding tests now. For instance, these two tests are currently passing. Is this what you mean?

```elixir
      assert to_from_ansidata(["Hello ", Owl.Data.tag("world", IO.ANSI.color(161)), "!"]) ==
               [["Hello ", Owl.Data.tag("world", "\e[38;5;161m")], "!"]

      assert to_from_ansidata([
               "Hello ",
               Owl.Data.tag("world", IO.ANSI.color_background(161)),
               "!"
             ]) == [["Hello ", Owl.Data.tag("world", "\e[48;5;161m")], "!"]
```

**Comment by zachallaun** [comment]
Rebased with `main`, added additional tests, and updated docs!

**Comment by fuelen** [comment]
Played a bit with it. Looks nice, good job.
Just noticed that it doesn't work with binaries, so getting data from cmd and trying to parse it won't work. Technically, `binary` is a part of `iodata`
```
> "test" |> Owl.Data.tag(:red) |> Owl.Data.to_ansidata() |> to_string() |> Owl.Data.from_ansidata()
"\e[31mtest\e[39m\e[0m"
```

**Comment by zachallaun** [comment]
> Just noticed that it doesn't work with binaries, so getting data from cmd and trying to parse it won't work. Technically, binary is a part of iodata

This is a documentation limitation for now, see [here](https://github.com/fuelen/owl/pull/11/files?diff=unified&w=0#diff-0d6d151c8319cc521966f92d519a82809ee2e18ac8bae8fa971191784b740808R306-R315).

It wouldn't be too hard to parse binaries to split out escape sequences in the future, but I chose not to add that for this first iteration. I agree that it would be nice to have in order to consume data from cmd. My personal use-case was focusing on consuming ansidata from `IO.ANSI`.

File: lib/owl/data/sequence.ex [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      type!(name)
+    end
+  end
+
+  @doc """
+  Get the sequence type of a sequence name.
+  """
+  def type!(name) do
```
[the remark this answers is not in the document]
Comment (zachallaun) [reply lib/owl/data/sequence.ex]: Are the changes in https://github.com/fuelen/owl/pull/11/commits/f0eb35f239886b0d590ff1d018775b2b073e6175 along the lines you were thinking?

[the remark this answers is not in the document]
Comment (fuelen) [reply lib/owl/data/sequence.ex]: yes! :)

**Comment by fuelen** [comment]
Yes, let's do binary parsing in a new iteration.

Thank you for the PR! :heart:

File: lib/owl/data.ex [hunk 5]
```diff
@@ -296,31 +298,87 @@ defmodule Owl.Data do
...
+        {tag_all(binary, open_tags), open_tags}
+
+      type ->
+        {[], update_open_tags(open_tags, type, Sequence.ansi_to_name(binary))}
+    end
   end
 
-  defp reverse_and_tag(sequences, [%Owl.Tag{sequences: last_sequences} | _] = data) do
-    maybe_wrap_to_tag(sequences -- last_sequences, Enum.reverse(data))
+  defp do_from_ansidata(integer, open_tags) when is_integer(integer) do
```
Comment (fuelen) [inline lib/owl/data.ex:346]: Looking at coverage and this line is red. This is for char list, right?

Comment (zachallaun) [reply lib/owl/data.ex:346]: Yeah, it’s just to handle individual character literals in ansidata. `[“hello”, \?]`

Should’ve added a test, sorry!
